src/persist-client/src/internal/machine.rs

```diff
@@ -343,21 +364,47 @@ where
         .await
     }
 
-    pub async fn heartbeat_reader(
+    pub async fn compare_and_downgrade_since<O: Codec64>(
         &mut self,
-        reader_id: &ReaderId,
+        reader_id: &CriticalReaderId,
+        expected_opaque: &O,
+        (new_opaque, new_since): (&O, &Antichain<T>),
+    ) -> (Result<Since<T>, (O, Since<T>)>, RoutineMaintenance) {
+        let metrics = Arc::clone(&self.metrics);
+        let (_seqno, res, maintenance) = self
+            .apply_unbatched_idempotent_cmd(
```

Comment:
how sure are we that this is idempotent? should probably write some tests here (somehow)

Reply:
I sort of have an idea on how to test this, but I'm not sure how clean it'll turn out. Will give it a shot